modules/programs/ssh.nix

```diff
@@ -15,6 +15,12 @@ let
 
   unwords = builtins.concatStringsSep " ";
 
+  mkSetEnvStr = envStr: unwords
+    (mapAttrsToList
+      (name: value: ''${name}="${value}"'')
```

Comment:
I think maybe the value needs to be escaped, e.g., in case it contains a `"` character?

Reply:
with `escapeShellArg` ?

Reply:
`(name: value: ''${name}="${escape [ "\"" ] value}"'')` ?

Reply:
Who knows. Read the source code, or experiment until you're confident with how the string is processed. Or just leave `value` raw and let the user figure it out.

Reply:
ssh -t checks the config (though it doesnt warn of everything), it's used in nixpkgs.

Reply:
I did some testing:
```
        setEnv.LC_EASY = "abc123";
        setEnv.LC_WITH_SPACE = "abc 123";
        setEnv.LC_WITH_1_QUOTE = ''abc" 123'';
        setEnv.LC_WITH_2_QUOTES = ''abc"" 123'';
```
I now have an implementation that works for all of the above test cases:
`(name: value: ''${name}="${escape [ "\"" ] value}"'')`

Result (on the remote host):
```
auriga% env | grep LC
LC_WITH_SPACE=abc 123
LC_WITH_1_QUOTE=abc" 123
LC_WITH_2_QUOTES=abc"" 123
LC_EASY=abc123
```

Reply:
How about `setEnv.LC_BACKSLASH = ''a\b'';`?

Reply:
> How about `setEnv.LC_BACKSLASH = ''a\b'';`?

It is rendered as this in the ssh config:
`  SetEnv LC_TEST="a\b"`

And, when evaluated on the host, it gives:
```
auriga% echo $LC_TEST 
a
```

I guess that it is the correct behavior... I mean that if the user wants to set `LC_FOO = ''a\b''`, he wants to have `SetEnv LC_FOO="a\b"` in his ssh config. What do you think ?

Reply:
If I set that I'd expect `LC_TEST` to contain `a\b` on the host. Try adding `\` to the list of escaped characters?

Reply:
Escaping it does produce `LC_TEST="a\\b"` in the ssh config but on the host it doesn't prevent `LC_TEST` from being just `a`.

More precisely, it seems that `LC_TEST` does contain the backslash but it cannot be "echo-ed"

Reply:
Glad `\"` works. I think it also makes sense to escape `\` since that is a pretty common character one might want to pass on. Don't think it's worth checking more characters, though. If somebody runs into any further ones then they can raise an issue and we'll fix it then.

Btw, please add a test case to [`tests/modules/programs/ssh`](https://github.com/nix-community/home-manager/tree/master/tests/modules/programs/ssh) for this to avoid regressions :-)

Reply:
Actually, the `home.sessionVariables` implementation does not escape anything:
https://github.com/nix-community/home-manager/blob/6abb775e75a7f25451781dc704cfe03f27ad7496/modules/lib/shell.nix#L5

Reply:
> In this regard, do you think that we should keep going with the escaping strategy or replicating the same thing as for `home.sessionVariables` (i.e. a simple `toString`) ?

@rycee @teto @ncfavier What is your opinion on this ?

Reply:
`sessionVariables` doesn't use `escapeShellArg` in order to allow variable expansion, but it should probably escape `"` and `\` to be honest.

In any case, I think this is fine; please add a test.

Reply:
Thank you for the feedback :)
I added a test.